Firestore/Source/Local/FSTMemoryPersistence.mm

```diff
@@ -105,4 +143,226 @@ - (void)shutdown {
+    FSTListenSequenceNumber sequenceNumber = it->second;
+    const DocumentKey &key = it->first;
+    if (![_persistence.queryCache containsKey:key]) {
+      block(key, sequenceNumber, &stop);
+    }
+  }
+}
+
+- (int)removeTargetsThroughSequenceNumber:(FSTListenSequenceNumber)sequenceNumber
+                              liveQueries:(NSDictionary<NSNumber *, FSTQueryData *> *)liveQueries {
+  return [_persistence.queryCache removeQueriesThroughSequenceNumber:sequenceNumber
+                                                         liveQueries:liveQueries];
+}
+
+- (int)removeOrphanedDocumentsThroughSequenceNumber:(FSTListenSequenceNumber)upperBound {
+  return [(FSTMemoryRemoteDocumentCache *)_persistence.remoteDocumentCache
+      removeOrphanedDocuments:self
+        throughSequenceNumber:upperBound];
+}
+
+- (void)addReference:(const DocumentKey &)key {
+  _sequenceNumbers[key] = self.currentSequenceNumber;
+}
+
+- (void)removeReference:(const DocumentKey &)key {
+  _sequenceNumbers[key] = self.currentSequenceNumber;
+}
+
+- (BOOL)mutationQueuesContainKey:(const DocumentKey &)key {
+  const MutationQueues &queues = [_persistence mutationQueues];
+  for (auto it = queues.begin(); it != queues.end(); ++it) {
+    if ([it->second containsKey:key]) {
+      return YES;
+    }
+  }
+  return NO;
+}
+
+- (void)removeMutationReference:(const DocumentKey &)key {
+  _sequenceNumbers[key] = self.currentSequenceNumber;
+}
+
+- (BOOL)isPinnedAtSequenceNumber:(FSTListenSequenceNumber)upperBound
+                        document:(const DocumentKey &)key {
+  if ([self mutationQueuesContainKey:key]) {
+    return YES;
+  }
+  if ([_additionalReferences containsKey:key]) {
+    return YES;
+  }
+  if ([_persistence.queryCache containsKey:key]) {
+    return YES;
+  }
+  auto it = _sequenceNumbers.find(key);
+  if (it != _sequenceNumbers.end() && it->second > upperBound) {
+    return YES;
+  }
+  return NO;
+}
+
+@end
+
+@implementation FSTMemoryEagerReferenceDelegate {
+  std::unique_ptr<std::unordered_set<DocumentKey, DocumentKeyHash> > _orphaned;
+  // This delegate should have the same lifetime as the persistence layer, but mark as
+  // weak to avoid retain cycle.
+  __weak FSTMemoryPersistence *_persistence;
+  FSTReferenceSet *_additionalReferences;
+}
+
+- (instancetype)initWithPersistence:(FSTMemoryPersistence *)persistence {
+  if (self = [super init]) {
+    _persistence = persistence;
+  }
+  return self;
+}
+
+- (FSTListenSequenceNumber)currentSequenceNumber {
+  return kFSTListenSequenceNumberInvalid;
+}
+
+- (void)addInMemoryPins:(FSTReferenceSet *)set {
+  // We should be able to assert that _additionalReferences is nil, but due to restarts in spec
+  // tests it would fail.
+  _additionalReferences = set;
+}
+
+- (void)removeTarget:(FSTQueryData *)queryData {
+  for (const DocumentKey &docKey :
+       [_persistence.queryCache matchingKeysForTargetID:queryData.targetID]) {
+    self->_orphaned->insert(docKey);
```

Comment:
Extraneous `self->`. (I only flag it because we're accessing `_orphaned` directly elsewhere in here.)

Reply:
Ah. It probably used to be in a block and then got migrated.